controller/destination/egress.go

```diff
@@ -0,0 +1,117 @@
+package destination
+
+import (
+	"fmt"
+	"net"
+	"sync"
+
+	common "github.com/runconduit/conduit/controller/gen/common"
+	log "github.com/sirupsen/logrus"
+)
+
+type DnsListener interface {
+	Update(add []common.TcpAddress, remove []common.TcpAddress)
+}
+
+type DnsWatcher struct {
+	hosts map[string]informer
+	mutex sync.Mutex
+}
+
+func NewDnsWatcher() *DnsWatcher {
+	return &DnsWatcher{
+		hosts: make(map[string]informer),
+	}
+}
+
+func (w *DnsWatcher) Subscribe(host string, listener DnsListener) error {
+	log.Printf("Establishing dns watch on host %s", host)
+
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+
+	informer, ok := w.hosts[host]
+	if !ok {
+		informer = newInformer(host)
+		w.hosts[host] = informer
+	}
+
+	informer.add(listener)
+
+	return nil
+}
+
+func (w *DnsWatcher) Unsubscribe(host string, listener DnsListener) error {
+	log.Printf("Stopping dns watch on host %s", host)
+
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+
+	informer, ok := w.hosts[host]
+	if !ok {
+		return fmt.Errorf("Cannot unsubscribe from %s: not subscribed", host)
+	}
+
+	informer.mutex.Lock()
+	defer informer.mutex.Unlock()
+
+	for l in range(informer.listeners) {
```

Comment:
if I remember correctly, this might be giving you the index instead of the item....